### PR TITLE
Add libc++ (fixes #6)

### LINKED ISSUE #6: Add the "libc++" C++ Standard Library to the build
[libc++](http://libcxx.llvm.org/) is a new implementation of the C++ standard library from the LLVM project, targeting C++11. It may be helpful to include it in the build for various reasons.

Here's an excerpt from libc++'s site:

> ### Why a new C++ Standard Library for C++11?
> 
> After its initial introduction, many people have asked "why start a new library instead of contributing to an existing library?" (like Apache's libstdcxx, GNU's libstdc++, STLport, etc). There are many contributing reasons, but some of the major ones are:
> - From years of experience (including having implemented the standard library before), we've learned many things about implementing the standard containers which require ABI breakage and fundamental changes to how they are implemented. For example, it is generally accepted that building std::string using the "short string optimization" instead of using Copy On Write (COW) is a superior approach for multicore machines (particularly in C++11, which has rvalue references). Breaking ABI compatibility with old versions of the library was determined to be critical to achieving the performance goals of libc++.
> - Mainline libstdc++ has switched to GPL3, a license which the developers of libc++ cannot use. libstdc++ 4.2 (the last GPL2 version) could be independently extended to support C++11, but this would be a fork of the codebase (which is often seen as worse for a project than starting a new independent one). Another problem with libstdc++ is that it is tightly integrated with G++ development, tending to be tied fairly closely to the matching version of G++.
> - STLport and the Apache libstdcxx library are two other popular candidates, but both lack C++11 support. Our experience (and the experience of libstdc++ developers) is that adding support for C++11 (in particular rvalue references and move-only types) requires changes to almost every class and function, essentially amounting to a rewrite. Faced with a rewrite, we decided to start from scratch and evaluate every design decision from first principles based on experience. Further, both projects are apparently abandoned: STLport 5.2.1 was released in Oct'08, and STDCXX 4.2.1 in May'08.

### PR DESCRIPTION
**Pull request by hesiod**
Add libc++-svn to the PKGBUILD, adapting some lines from the libc++ PKGBUILD,
to be found at https://aur.archlinux.org/cgit/aur.git/tree/PKGBUILD?h=libc%2B%2B.

I still have to test this a bit more, but it should generally work.

### CODE REVIEW

**Comment by brunocodutra** [comment]
Awesome!

**Comment by kerberizer** [comment]
@hesiod Thank you very much for the pull request. I know libc++ is important for some people, but unfortunately I've never been able to invest more time in acquainting myself better with it.

Here are some points that we'll probably need to take care of (starting from the least important ones):
* Could you please set the indentation in the two new package functions to 4 spaces?
* I think it'll be best if we leave those {LICENSE,CREDITS}.TXT files as they are. BTW, thanks for noticing the CREDITS one in the first place, as it seems to be needed at least also for compiler-rt. I'll actually change `_install_license()` to install both files (or whichever of them exists) and from the correct directory, so we may use it directly in the two new package functions as well.
* I see that some unit tests are failing. More on this little further below.

My biggest concern so far are the [instructions](http://libcxx.llvm.org/docs/BuildingLibcxx.html) suggesting that Clang is the preferred compiler. I wonder if we shouldn't indeed make the additional effort to use it instead of GCC. The obvious problem is which Clang should we use. There are several possible approaches (unless I'm missing something of course):
* Just use GCC. This obviously is the easiest thing, but I'm worried if we wouldn't get into some compatibility problems. Indeed, when the library is built with GCC, four of the libcxx unit tests fail. These problems actually worry me more than if the compilation was failing right away, as they can lead to some difficult to debug user problems.
* Add Clang to `makedepends`. This would pull Clang from `extra` on a clean system (including the chroots) or use the clang-svn that may be already installed (we may also explicitly ask for `clang-svn`, but this means forcing the users to use the binary repo or to build in two stages). With the Clang from `extra` only one unit test fails, so this is kind of an improvement. I've actually tested building everything with Clang, although it might be possible to build only libc++/abi with it.
* Use the freshly built Clang. This probably is the best approach overall, but also the hardest to achieve. Essentially, we should do the [two-stage bootstrap build](http://llvm.org/docs/AdvancedBuilds.html), which would also bring the additional advantage of having Clang bootstrapped in a more traditional way (if we really want to go hardcore, we could even do the three-stage bootstrap). I'm not sure however whether libc++ would fit naturally into this workflow or it'll require some hacking.

Being a perfectionist, I'm leaning towards the last variant, but again, it might require some non-trivial effort, and so it might be better to leave it as a todo. I'll need to do more testing with different setups, but so far my feelings are that we should avoid GCC if possible and pulling Clang might be the most balanced short-term approach. And we definitely need to have all unit tests passing, or at least know why they fail and that it's not a serious concern (like what's happening with those AMDGPU tests in #12).

Once again, really appreciate your help and will be glad to hear your opinion.

_Addendum: the unit tests that fail._

**When built with Clang**
```
Failing Tests (1):
    libc++ :: std/atomics/atomics.lockfree/isalwayslockfree.pass.cpp
```

**When built with GCC**
```
Failing Tests (4):
    libc++ :: std/depr/depr.c.headers/math_h.pass.cpp
    libc++ :: std/numerics/c.math/cmath.pass.cpp
    libc++ :: std/numerics/rand/rand.util/rand.util.canonical/generate_canonical.pass.cpp
    libc++ :: std/utilities/time/time.duration/time.duration.nonmember/op_divide_duration.pass.cpp
```

**Comment by kerberizer** [comment]
I've updated the patch, but the web editor produced a merged commit that'll probably make the history a bit hard to follow. I'll look into this again when we decide to merge back to master: might simply cherry-pick or custom rebase the needed commits, a bit like I did with #9. Let's keep it the way it is for reference for now.
